crates/bevy_ecs/src/query/state.rs

```diff
@@ -24,26 +24,101 @@ use super::{
-/// a [`QueryState`] will be all [`TableId`]s or all [`ArchetypeId`]s, and not a mixture of both. This
-/// removes the need for discriminator to minimize memory usage and branching during iteration, but requires
-/// a safety invariant be verified when disambiguating them.
-///
-/// # Safety
-/// Must be initialized and accessed as a [`TableId`], if both generic parameters to the query are dense.
-/// Must be initialized and accessed as an [`ArchetypeId`] otherwise.
-#[derive(Clone, Copy)]
-pub(super) union StorageId {
-    pub(super) table_id: TableId,
-    pub(super) archetype_id: ArchetypeId,
+#[derive(Copy, Clone)]
+pub(super) enum StorageId {
+    Archetype(ArchetypeId),
+    Table(TableId),
+}
+
+impl StorageId {
+    fn as_archetype_id(&self) -> Option<ArchetypeId> {
+        match self {
+            Self::Archetype(id) => Some(*id),
+            _ => None,
+        }
+    }
+
+    /// # Safety
+    ///
+    /// Needs a safety comment before merging!
+    pub(super) unsafe fn debug_checked_as_archetype_id(&self) -> ArchetypeId {
+        self.as_archetype_id().debug_checked_unwrap()
+    }
+
+    fn as_table_id(&self) -> Option<TableId> {
+        match self {
+            Self::Table(id) => Some(*id),
+            _ => None,
+        }
+    }
+
+    /// # Safety
+    ///
+    /// Needs a safety comment before merging!
+    pub(super) unsafe fn debug_checked_as_table_id(&self) -> TableId {
+        self.as_table_id().debug_checked_unwrap()
+    }
+}
+
+impl From<TableId> for StorageId {
+    fn from(value: TableId) -> Self {
+        Self::Table(value)
+    }
+}
+
+impl From<ArchetypeId> for StorageId {
+    fn from(value: ArchetypeId) -> Self {
+        Self::Archetype(value)
+    }
+}
+
+pub(super) enum StorageIdIter<'s> {
+    Archetypes(core::slice::Iter<'s, ArchetypeId>),
+    Tables(core::slice::Iter<'s, TableId>),
+}
+
+impl Iterator for StorageIdIter<'_> {
+    type Item = StorageId;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::Tables(table_ids) => table_ids.next().copied().map(Into::into),
+            Self::Archetypes(archetype_ids) => archetype_ids.next().copied().map(Into::into),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(super) enum StorageIds {
+    Archetypes(Vec<ArchetypeId>),
+    Tables(Vec<TableId>),
+}
+
+impl StorageIds {
+    fn new_from_query_types<D: QueryData, F: QueryFilter>() -> Self {
+        Self::new_from_dense_flag(D::IS_DENSE && F::IS_DENSE)
+    }
+
+    fn new_from_dense_flag(dense: bool) -> Self {
+        match dense {
+            true => Self::Tables(Vec::new()),
+            false => Self::Archetypes(Vec::new()),
+        }
+    }
+
+    fn is_tables(&self) -> bool {
+        matches!(self, Self::Tables(_))
+    }
+
+    #[cfg(all(not(target_arch = "wasm32"), feature = "multi_threaded"))]
+    fn iter(&self) -> StorageIdIter<'_> {
+        match self {
+            Self::Tables(table_ids) => StorageIdIter::Tables(table_ids.iter().clone()),
```

Comment:
Do you need the two `clone()` calls in this method?  It seems like you could use the original `slice::Iter` values directly.

Reply:
I think you're right! I think this was a bit of a copy&paste error from my other `iter` implementation where I was cloning existing iterators so as to not consume them. The tests pass with this removed 🎉